content/guides/advanced/testing-with-jest-and-bankrun.md

```diff
@@ -0,0 +1,380 @@
+
+```json
+{
+  "scripts": {
+    "test": "jest"
+  }
+}
+```
+
+Then you can run the tests with:
+
+```bash
+yarn test
+```
+
+Since we want to run our tests with Typescript we need to install the `ts-jest`
+package and also create a Jest configuration file:
+
+```bash
+yarn add --dev ts-jest @jest/globals @types/jest
+yarn ts-jest config:init
+```
+
+This will create a `jest.config.js` file in your project. Now you can update
+your `anchor.toml` file to run the Jest tests:
+
+```bash
+test = "yarn test"
+```
+
+### Jest Troubleshooting
+
+1. In case your get a
+   `SyntaxError: Cannot use import statement outside a module` error you either
+   did not create a jest config or you need to add the following to your
+   `jest.config.js` file:
+
+```js
+module.exports = {
+  transform: {
+    "^.+\\.tsx?$": "ts-jest",
+  },
+  testEnvironment: "node",
+  moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
+};
+```
+
+2. Since running tests against the local validator can take quite some time, if
+   you get an error saying that you can not log after the test is finished your
+   test is probably timing out. Different to Mocha Jest does not have a default
+   timeout. You can set a timeout in your `jest.config.js` file:
+
+```js
+module.exports = {
+  testTimeout: 10000,
+};
+```
+
+Or you can set a timeout for a single test:
+
+```js
+test("test name", async () => {
+  // your test code
+}, 10000);
+```
+
+3. If you get an error saying that Anchor environment variables are missing you
+   are probably trying to use the AnchorProvider without running the tests
+   through `anchor test` or `anchor run test`. Just update your Anchor.toml to
+   run the Jest tests `yarn test` to run them in the anchor environment with all
+   the environment variables set.
+
+## Bankrun
+
+Instead of using `solana-test-validator` you can also use
+[Solana Bankrun](https://kevinheavey.github.io/solana-bankrun/). It acts
+similarly to the local validator but is more lightweight and faster. Some really
+helpful features are
+[writing custom account data](https://kevinheavey.github.io/solana-bankrun/tutorial/#writing-arbitrary-accounts)
+and
+[time travel](https://kevinheavey.github.io/solana-bankrun/tutorial/#time-travel)
+which makes it much easier to test time based programs and programs that rely on
+specific account data.
+
+To use Bankrun and Bankrun Anchor you need to add it to your `package.json`:
+
+```bash
+yarn add solana-bankrun anchor-bankrun
+```
+
+To switch from a Mocha Anchor test to a
+[Bankrun Anchor](https://github.com/kevinheavey/anchor-bankrun) test you only
+need to change the provider to be a
+[`BankrunProvider`](https://kevinheavey.github.io/solana-bankrun/tutorial/#anchor-integration)
+and create a context using `startAnchor`:
+
+```js
+const context = await startAnchor(".", [], []);
+const client = context.banksClient;
+
+const provider = new BankrunProvider(context);
+anchor.setProvider(provider);
+```
```

Comment:
can you add some more context into this code snippet of where these varaibles should go?

maybe a code snippet that looks resembles this (idk if this is exactly right, so please adjust to make it correct):

````
```js filename="my-test.test.ts" /startAnchor/
// ... imports here

describe('My test', async () => {
  const context = await startAnchor(".", [], []);
  const client = context.banksClient;

  const provider = new BankrunProvider(context);
  anchor.setProvider(provider);

  test('Test allocate counter + increment tx', async () => {
    // ... testing logic
  }
}
```
````